src/SteadyHeap.h

```diff
@@ -0,0 +1,107 @@
+/// complete interface for a generic heap, and the implementation is expected to
+/// be extended when new requirements arise in the library.
+///
+/// \tparam T The type of data the heap will contain as its elements.
+/// \tparam StrictOrdering The type of the comparison operation that will be
+/// used to determine the positioning of elements in the heap. It should satisfy
+/// the requirements of _Compare_, meaning that the return value of the
+/// comparison function, when contextually converted to `bool`, should yield
+/// `true` if its first argument appears before the second in the strict weak
+/// ordering induced by the type of its parameters. Comparison using less-than
+/// yields a max heap, and greater-than yields a min heap.
+template <typename T, typename StrictOrdering>
+class SteadyHeap {
+ private:
+  /// The fixed-size space in memory that the heap is realized in.
+  std::vector<T> _heap;
+  /// The strict ordering that is used to compare elements inside the heap.
+  StrictOrdering _ordering;
+
+ public:
+  /// Construct a SteadyHeap with a strict ordering.
+  ///
+  /// \param ordering See the template parameter StrictOrdering for more
+  /// information. Probably best implemented as a std::function wrapper.
+  SteadyHeap(StrictOrdering ordering) : _ordering(ordering) {}
+
+  /// \return Whether the container is empty.
+  bool empty() const { return _heap.empty(); }
+
+  /// \return The number of elements in the container.
+  std::size_t size() const { return _heap.size(); }
+
+  /// \return The reserved size of the container.
+  std::size_t capacity() const { return _heap.capacity(); }
+
+  /// \return An unmodifiable reference to the element at the top of the heap.
+  const T& front() const { return _heap.front(); }
+
+  /// Reserves the space in memory the heap is allowed to occupy at its fullest.
+  void reserve(std::size_t n) { _heap.reserve(n); }
+
+  /// Finds an element in the heap using `std::find`.
+  typename std::vector<T>::iterator find(const T& value) {
+    return std::find(_heap.begin(), _heap.end(), value);
+  }
+
+  /// Counts the number of occurrences of an element using `std::count`.
+  std::ptrdiff_t count(const T& value) const {
+    return std::count(_heap.begin(), _heap.end(), value);
+  }
+
+  /// Push an element onto the heap.
+  void push_heap(T& value) {
+    _heap.push_back(value);
+    std::push_heap(_heap.begin(), _heap.end(), _ordering);
+  }
+
+  /// Pop the element at the top of the heap.
+  void pop_heap() {
+    std::pop_heap(_heap.begin(), _heap.end(), _ordering);
+    _heap.pop_back();
+  }
+
+  /// A special method that forces an element onto the top of the heap, and then
+  /// pops it from there. This is useful when a modified element has to be
+  /// removed from (and later reinserted into) the heap, because the
+  /// modification could have affected its position inside the heap.
+  void force_pop_heap(const T& value) {
```

Comment:
This is a nasty one. Because the tricky here is that in order to find the value, we are using equality. If we do this, it is impossible that the ordering has changed. The trick is that the ordering functions are not using the ordering of these values, but rather these values are IDs and the ordering uses something behind the scenes. I don't know how could we show this special behavior in the naming. My original idea to name this "modify_heap" would also be misleading in this regard. So I honestly don't know if there is a better naming for this or not.

Reply:
I think this operation should indeed keep the heap property, it just forces an element out of the container. The key here might be that this will not pop according to the heap rules, so the semantics of the removal lies outside of the bounds of any heap logic, and maybe having "force" in the name is enough to signify that.

Reply:
Ok, this is logical. Maybe doc can say that it is working by using equality.

Reply:
Well, I can't see what else this could be misunderstood as. It can only find the element using equality, so we might say that the nature of the function already explains that, and I think that might be enough.